_posts/2019-11-19-demo-fraud-detection.md

```diff
@@ -0,0 +1,223 @@
+
+The demo out of the box comes with a set of predefined sample rules. You can
+click the _Start_ button and after some time will observe alerts displayed in the right section of the UI. These alerts are the result of Flink evaluating the generated transactions stream against the predefined rules.
+
+ Our sample fraud detection system consists of three main components:
+
+  1. Frontend (React)  
+  1. Backend (SpringBoot)  
+  1. Fraud Detection application (Apache Flink)  
+
+Interactions between the main elements are depicted in _Figure 2_.
+
+ <center>
+ <img src="{{ site.baseurl }}/img/blog/2019-11-19-demo-fraud-detection/architecture.png" width="800px" alt="Figure 2: Demo Components"/>
+ <br/>
+ <i><small>Figure 2: Fraud Detection Demo Components</small></i>
+ </center>
+ <br/>
+
+ Backend exposes a REST API to the Frontend for creating and deleting rules as well as issuing control commands for managing the demo execution. It then relays those Frontend actions to Flink by sending them via a "Control" Kafka topic. Backend also includes a _Transactions Generator_ component, which sends an emulated stream of money transfer events to Flink via a separate "Transactions" topic. Alerts generated by Flink are consumed by the Backend from "Alerts" topic and relayed to the UI via WebSockets.
+
+Having the familiarity with the overall layout and the goal of our Fraud Detection engine, let's now go into the details of what is required to implement such system.
+
+### Dynamic Data Partitioning
+
+The first pattern we will look into is Dynamic Data Partitioning.
+
+Flink's DataStream API includes an important operator that determines the distribution of incoming data (shuffle) among the parallel computational nodes of the cluster - the **keyBy()** operator. It partitions the stream in such a way that all records with the same key are assigned to the same partition and are hence sent to the same physical instance of the subsequent "keyed" operator.
+
+In a typical statically-defined streaming application the choice of the key is fixed and it is usually immediately derived from one of the fields of the incoming records. For instance, in case of a simple window-based computation over a stream of Tuples, we could use the first field of type String (denoted by 0) as the key:
+
+```java
+DataStream<Tuple2<String, Integer>> input = // [...]
+DataStream<...> windowed = input
+  .keyBy(0)
+  .window(/*window specification*/);
+```
+
+This approach is the basic building block for achieving horizontal scalability in a wide range of use cases. However, in the case of an application that strives to provide flexibility in business logic at runtime, this is not enough.
```

Comment:
```suggestion
This approach is one of the building blocks for achieving horizontal scalability in a wide range of use cases. However, in the case of an application striving for flexibility in business logic at runtime, this is not enough.
```

Reply:
I guess I wanted to say that this the most important one. "This approach is the main building block" ?

Reply:
makes sense. Basic sounded a bit limiting (like it's the least that needs to be done). 
I updated the suggestion with "main"